receiver/awscontainerinsightreceiver/internal/cadvisor/cadvisor_linux.go

```
@@ -17,24 +17,227 @@
+	perfEventsFile string) (cadvisorManager, error) {
+	return manager.New(memoryCache, sysfs, houskeepingConfig, includedMetricsSet, collectorHTTPClient, rawContainerCgroupPathPrefixWhiteList, perfEventsFile)
+}
+
+type cadvisorOption func(*Cadvisor)
+
+func cadvisorManagerCreator(f createCadvisorManager) cadvisorOption {
+	return func(c *Cadvisor) {
+		c.createCadvisorManager = f
+	}
+}
+
+type hostInfo interface {
+	GetNumCores() int64
+	GetMemoryCapacity() int64
+	GetClusterName() string
+}
+
 type Cadvisor struct {
-	// TODO: add proper field for Cadvisor
+	logger                *zap.Logger
+	nodeName              string //get the value from downward API
+	createCadvisorManager createCadvisorManager
+	manager               cadvisorManager
+	version               string
+	hostInfo              hostInfo
+	containerOrchestrator string
+}
+
+func overrideCadvisorFlagDefault(logger *zap.Logger) {
+	flagOverrides := map[string]string{
+		// Override the default cAdvisor housekeeping interval.
+		"housekeeping_interval": defaultHousekeepingInterval.String(),
+		// override other defaults (in future)
+	}
+	for name, defaultValue := range flagOverrides {
+		if f := flag.Lookup(name); f != nil {
+			f.DefValue = defaultValue
+			f.Value.Set(defaultValue)
+		} else {
+			logger.Warn("Expected cAdvisor flag not found", zap.String("name", name))
+		}
+	}
 }
 
 // New creates a Cadvisor struct which can generate metrics from embedded cadvisor lib
-func New(containerOrchestrator string, machineInfo *host.MachineInfo, logger *zap.Logger) *Cadvisor {
-	// TODO: initialize the cadvisor
-	return &Cadvisor{}
+func New(containerOrchestrator string, hostInfo hostInfo, logger *zap.Logger, options ...cadvisorOption) *Cadvisor {
+	nodeName := os.Getenv("HOST_NAME")
+	if nodeName == "" {
+		err := errors.New("missing environment variable HOST_NAME. Please check your deployment YAML config")
```

Can we return the err?

---

Sure. I will make the change.